### ISSUE TITLE
[BUG] Editing translations of pages using the language tabs inside PageContent creates an inconsistent state 

### ISSUE THREAD

**Issue by filipweidemann** [django-cms/django-cms#8044]
## Description

Consider the following scenario: we got a multi-language site using the languages English (en) and German (de). 

We currently got a page which only exists in the German (de) version. Therefore, we only have a published `PageContent` object for this page in German.

If we change the language switch inside the `PageContent` admin list view (the one on top, just beside the "Main Navigation" label) to `English (en)`, we can see the title of the German Page with a "(de)" behind it. We can then click on the grey circle to create the english version of the page. This all works flawlessly.

However, suppose we have a page which is published in both versions. We now click on the page content settings on the German version and then use the language tab on top of the form to switch the language to English.
This results in an empty form without title, slug or anything that the *already existing english version of this page* provides. If we continue to fill out the form and save it, we can create an inconsistent state inside the DB because the `PageUrl` objects now return 2 results for the english page version. 

If we look at the `get_urls()` method of this page, we can see that the English version now has 2 PageUrl objects:

`<QuerySet [<PageUrl: english-testpage (en)>, <PageUrl: deutsche-testseite-translated (de)>, <PageUrl: adskjfhgdskjfhdsakjh (en)>]>`

If we then select the language with 2 `PageUrl` objects in the `PageContent` list view and try to edit it, we get an error:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/contrib/admin/options.py", line 688, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/utils/decorators.py", line 134, in _wrapper_view
    response = view_func(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/views/decorators/cache.py", line 62, in _wrapper_view_func
    response = view_func(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/contrib/admin/sites.py", line 242, in inner
    return view(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/cms/admin/pageadmin.py", line 981, in change_view
    return super().change_view(request, object_id, form_url=form_url, extra_context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/contrib/admin/options.py", line 1889, in change_view
    return self.changeform_view(request, object_id, form_url, extra_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/utils/decorators.py", line 46, in _wrapper
    return bound_method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/utils/decorators.py", line 134, in _wrapper_view
    response = view_func(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/contrib/admin/options.py", line 1747, in changeform_view
    return self._changeform_view(request, object_id, form_url, extra_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/contrib/admin/options.py", line 1819, in _changeform_view
    form = ModelForm(instance=obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/cms/admin/forms.py", line 578, in __init__
    self.url_obj = self.instance.page.get_url(self._language)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/cms/models/pagemodel.py", line 782, in get_url
    return self.get_urls().get(language=language)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/django/db/models/query.py", line 640, in get
    raise self.model.MultipleObjectsReturned(
    ^

Exception Type: MultipleObjectsReturned at /de/admin/cms/pagecontent/362/change/
Exception Value: get() returned more than one PageUrl -- it returned 2!
```

This is really unexpected because we get an explicit option to edit page contents in more than one way and depending on which way you choose you may end up destroying the integrity of your pages.

## Steps to reproduce

Follow the steps above.

## Expected behaviour

Page Content always get correctly fetched and edited without destroying the consistency of the DB

## Actual behaviour

See above.

## Screenshots

Not really applicable, the reproduction is too complex to convey through screenshots.

## Additional information (CMS/Python/Django versions)
Python 3.11
Django 4.2
djangoCMS 4.1.3

## Do you want to help fix this issue?

Imo, this can only be resolved in two ways: 
1) fix the edit view of page content objects so that using the tab select correctly fetches existing page information and edits it as expected (just the way it did in djangoCMS 3) or
2) completely disable the tab selectors inside the page content edit view so that you NEED to use the language select inside the list view.

I know this may seem hard to fathom and especially hard to reproduce, so if anyone needs clarification or a short demo I am always available to hop on a quick Discord call.

So yes, I'd like to help but I can't really dig through djangoCMS internals right now and create a PR. Always happy to review them by testing however :)

* [x] Yes, I want to help fix this issue and I will join the channel #pr-reviews on [the Discord Server](https://discord-pr-review-channel.django-cms.org) to confirm with the community that a PR is welcome.
* [ ] No, I only want to report the issue.

**Comment by filipweidemann** [comment]
Further investigation: 
there are already some issues regarding this (or parts of this) issues, e.g.
https://github.com/django-cms/django-cms/issues/7838
https://github.com/django-cms/django-cms/issues/7866

It seems like just removing the language tabs would at least prevent the underlying issues from ever getting created by user input.

**Comment by fsbraun** [comment]
@filipweidemann Great catch! Thanks for reporting this.

**Comment by fsbraun** [comment]
@filipweidemann I just tested with the above fix and it avoids the "illegal" creation of double entries in the url table. But it still unexpectedly does create a new version without one wanting it.

**Comment by filipweidemann** [comment]
@fsbraun what a timing, I was typing something just now as well :D

I already tested your PR earlier today and it didn't quite fix it (just like you said).
I'm currently peeking into the codebase and I suspect that the issue is somewhere inside this template tag:
https://github.com/django-cms/django-cms/blob/develop-4/cms/templatetags/cms_admin.py#L36

It gets called by the template to generate the links of each "language tab" and if I bootstrap a fresh and minimal djangoCMS installation with two languages, this method *always* uses a "create" link for the unselected tab rather than fetching the content and displaying a "change" link so that I actually jump into the edit form of each languages' `PageContent`.

I also verified that sometimes the underlying `get_admin_content` method of the `Page` model returns an `EmptyPageContent` instance. Looking into the definition of this class, we can spot the `__bool__` method that always returns false. 

So, whenever an `EmptyPageContent` instance gets returned, the template tags `if page_content` clause (https://github.com/django-cms/django-cms/blob/develop-4/cms/templatetags/cms_admin.py#L39) never triggers, sending us into the path that creates a "create" link for the languages' tab.

Now, the only question is *why exactly* the method `get_admin_content` sometimes returns an `EmptyPageContent` rather than the latest available `PageContent`. I suspect that there is an issue with the way the cache of a Page instance gets created.

I can keep digging later tonight, but this should already be quite helpful.

**Comment by filipweidemann** [comment]
Yup, I am not sure if the actual problem isn't somewhere inside djangocms-versioning but I can say for sure that the issue is the following:

Whenever I access the page edit view, I get thrown into the correct language, the one I selected in the list views' dropdown menu. The method `get_admin_content` gets called by the template tag to determine what URL **THE OTHER** button should point to, and there is a check that calls `set_admin_content_cache`. This method does the correct thing, it populates the cache with all language versions. 
However, it only gets called if the cache is not already populated with **something**. In this case, it is. 
To be precise, it will be populated with **one and only one** entry, the one of the current page. That means that the `self.admin_content_cache.get(language)` will fail because the key for the DE language is not present inside the cache, returning the default `EmptyPageContent`, evaluating to `False` inside the template tag because of its `__bool__` implementation, returning a URL that does not point to an edit, but a create form. Which is exactly the issue.

**Comment by fsbraun** [comment]
@filipweidemann **Excellent work!** I have not looked into it, but this makes sense:

* `get_admin_content` was introduced in #8002 and its cache needs to be either set or not set for all languages
* `set_admin_content` does always add all missing languages
* `cms.admin.pageadmin.get_object`, however, sets the cache for one object only. This is unnecessary - and I think also wrong.

Can you try removing those two lines? https://github.com/django-cms/django-cms/blob/bb705e629a20b482bf02511924ac8cc4011f9f69/cms/admin/pageadmin.py#L744-L746

It works for me. I have to ponder if that has any performance implications. How does it look from your side? Would you like to create a PR?

**Comment by filipweidemann** [comment]
I couldn't really find the cause why the cache is not completely populated but my hunch above is correct. When removing the if-check that stops the cache from always being recreated (https://github.com/django-cms/django-cms/blob/bb705e629a20b482bf02511924ac8cc4011f9f69/cms/models/pagemodel.py#L731) everything just starts working.

I'll stop spamming half-a**ed solutions now but the cache being present but not populated for all available languages is the main culprit here, I am 100% certain that fixing whatever causes the weird cache behaviour also fixes the issue inside the admin.

Lmk if I can assist further :)

**Comment by fsbraun** [comment]
Just filling the cache every time will cause an N + 1 problem. I'd be very happy if you tried removing those two bad lines from `pageadmin.py`. 

Also I think https://github.com/django-cms/django-cms/blob/bb705e629a20b482bf02511924ac8cc4011f9f69/cms/models/pagemodel.py#L724-L726 might need an update to
```
    def set_admin_content_cache(self):
        for translation in self.pagecontent_set(manager="admin_manager").latest_content().all():
            self.admin_content_cache.setdefault(translation.language, translation)
```
The difference is that `current_content` only gets draft and published versions (which is ok in most of the cases), but if you  only had an unpublished version, you'd also like to see their settings.

**Comment by filipweidemann** [comment]
Today was insanely rough and I am quite done, but I'll create a PR tomorrow! I am not sure if I understand the whole picture yet, let me think about the stuff you wrote and I'll try to jump through the code to make sure I fully understand everything before creating a PR.

**Comment by filipweidemann** [comment]
@fsbraun I understand the whole thing now. Setting the cache inside the `get_object` method only does so for the currently active language, meaning that using the cache later on (in this case, for getting the language links) results in the observed behaviour. 

I removed the lines responsible for setting the cache inside the `get_object` wrapper like you suggested and it fixes the issue because the admin cache is now completely empty and not only partially filled, meaning that we set it up when calling the `get_admin_content` method.

As for your performance considerations: I am quite sure that this is okay. We only fetch `PageContent` objects per additional language & do so using literally *one* query. This should not be a problem at all I think.
Also, this only is done when using the admin, the final webpage does not do any of this.

Here's the PR, please have a look and I would be glad for a fast merge if possible :D
https://github.com/django-cms/django-cms/pull/8046

**Comment by fsbraun** [comment]
@marksweb I do not see a direct test for this PR, but I'd like to introduce a `CacheDict` class (subclass of `dict`) which has `__setitem__` raise an `AttributeError` to avoid people meddling with the cache in the future.

Now, this bug turns out to be a left-over of V3, but the `admin_content_cache` is either filled or not, and must not be partially filled. Overwriting `__setitem__` might prevent people from directly setting items.

What do you think?

**Comment by marksweb** [comment]
@fsbraun if this relates to that PR where I mentioned tests then that sounds great. I was essentially looking for something that allows us to avoid regression.

**Comment by fsbraun** [comment]
Fixed by #8049 and #8046
